src/main/java/org/spongepowered/common/conversation/SpongeConversation.java

```diff
@@ -0,0 +1,219 @@
+import org.spongepowered.api.plugin.PluginContainer;
+import org.spongepowered.common.SpongeImpl;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.Nullable;
+
+public class SpongeConversation implements Conversation {
+
+    private final PluginContainer creator;
+    private final ConversationArchetype archetype;
+
+    private final Map<Conversant, ExternalChatHandler> externalChatHandlers = new ConcurrentHashMap<>();
+    private final DataContainer context = new MemoryDataContainer();
+    private final Set<EndingHandler> endingHandlers;
+    @Nullable private Question currentQuestion;
+    private boolean catchesOutput;
+    private boolean allowCommands;
+
+    private boolean ended;
+
+    /**
+     * Creates a new conversation.
+     *
+     * @param archetype The archetype for the question
+     * @param conversants All conversants in the conversation
+     * @param creator Who created the conversation
+     */
+    SpongeConversation(ConversationArchetype archetype, Collection<Conversant> conversants, PluginContainer creator) {
+        this.archetype = archetype;
+        conversants.forEach(c -> addConversant(c, archetype.getDefaultChatHandler()));
+        this.creator = creator;
+        this.endingHandlers = new HashSet<>(archetype.getEndingHandlers());
+        this.catchesOutput = archetype.catchesOutput();
+        this.allowCommands = archetype.allowsCommands();
+        this.currentQuestion = archetype.getFirstQuestion();
+    }
+
+    @Override
+    public DataContainer getContext() {
+        return this.context;
+    }
+
+    @Override
+    public PluginContainer getCreator() {
+        return this.creator;
+    }
+
+    @Override
+    public ConversationArchetype getArchetype() {
+        return this.archetype;
+    }
+
+    @Override
+    public Set<Conversant> getConversants() {
+        return this.externalChatHandlers.keySet();
+    }
+
+    @Override
+    public Optional<Question> getCurrentQuestion() {
+        return Optional.ofNullable(this.currentQuestion);
+    }
+
+    @Override
+    public void setQuestion(@Nullable Question question) {
+        this.currentQuestion = question;
+        if (question != null) {
+            this.archetype.getHeader().ifPresent(text -> getConversants().forEach(c -> c.sendThroughMessage(text)));
+            getConversants().forEach(c -> c.sendThroughMessage(question.getPrompt()));
+        }
+    }
+
+    @Override
+    public boolean end(ConversationEndType endType, Cause cause) {
+        cause = cause.with(NamedCause.of(endType.getName(), ConversationEndCause.builder().type(endType).build()));
+        final ConversationCloseEvent.Ending endingEvent = SpongeEventFactory.createConversationCloseEventEnding(cause, this);
+
+        if (SpongeImpl.postEvent(endingEvent) || endingEvent.isCancelled()) {
```

Comment:
`SpongeImpl.postEvent` returns whether the event was (not?) cancelled so the extra check isn't necessary.